new-new-packages/compiler/src/compiler/string_to_rcst.rs

```diff
@@ -0,0 +1,1394 @@
+use super::rcst::{Rcst, RcstError};
+use crate::input::{Input, InputDb};
+use std::sync::Arc;
+
+#[salsa::query_group(StringToRcstStorage)]
+pub trait StringToRcst: InputDb {
+    fn rcst(&self, input: Input) -> Option<Arc<Vec<Rcst>>>;
+}
+
+fn rcst(db: &dyn StringToRcst, input: Input) -> Option<Arc<Vec<Rcst>>> {
+    let source = db.get_input(input)?;
+    let (rest, mut rcsts) = parse::body(&source, 0);
+    if !rest.is_empty() {
+        rcsts.push(Rcst::Error {
+            unparsable_input: rest.to_string(),
+            error: RcstError::UnparsedRest,
+        });
+    }
+    Some(Arc::new(rcsts))
+}
+
+impl Rcst {
+    fn wrap_in_whitespace(mut self, mut whitespace: Vec<Rcst>) -> Self {
+        if !whitespace.is_empty() {
+            if let Rcst::TrailingWhitespace {
+                whitespace: self_whitespace,
+                ..
+            } = &mut self
+            {
+                self_whitespace.append(&mut whitespace);
+                self
+            } else {
+                Rcst::TrailingWhitespace {
+                    child: Box::new(self),
+                    whitespace,
+                }
+            }
+        } else {
+            self
+        }
```

Comment:
Making this a guard clause improves readability:

```suggestion
        if whitespace.is_empty() {
            return self;
        }

        if let Rcst::TrailingWhitespace {
            whitespace: self_whitespace,
            ..
        } = &mut self
        {
            self_whitespace.append(&mut whitespace);
            self
        } else {
            Rcst::TrailingWhitespace {
                child: Box::new(self),
                whitespace,
            }
        }
```